src/main/java/org/reaktivity/nukleus/http2/internal/routable/stream/TargetOutputEstablishedStreamFactory.java

```diff
@@ -277,24 +274,24 @@ private void processData(
 
             if (extension.sizeof() > 0)
             {
+                int promisedStreamId = promisedStreamIds.getAsInt();
                 Http2DataExFW dataEx = extension.get(dataExRO::wrap);
-                promisedStreamId += 2;
-                pushPromiseRW
+                Http2PushPromiseFW pushPromise = pushPromiseRW
                         .wrap(writeBuffer, 0, writeBuffer.capacity())
                         .streamId(http2StreamId)
                         .promisedStreamId(promisedStreamId)
-                        .endHeaders();
-                dataEx.headers().forEach(httpHeader -> mapPushHeader(hpackContext, httpHeader));
-                Map<String, String> promisedHeaders = new HashMap<>();
-                dataEx.headers().forEach(
-                        httpHeader -> promisedHeaders.put(httpHeader.name().asString(), httpHeader.value().asString()));
-                Http2PushPromiseFW pushPromise = pushPromiseRW.build();
+                        .endHeaders()
+                        .set(dataEx.headers(), mapHeader(hpackContext))
+                        .build();
+
                 // TODO remove the following and throttle based on HTTP2_WINDOW update
                 target.addThrottle(sourceOutputEstId, this::handleThrottle);
                 target.doData(sourceOutputEstId, pushPromise.buffer(), pushPromise.offset(), pushPromise.limit());
-                // TODO create and add Http2Stream for promised stream
 
-                session.doPromisedRequest(promisedStreamId, promisedHeaders);
+                Map<String, String> promisedHeaders = new HashMap<>();
+                dataEx.headers().forEach(
+                        httpHeader -> promisedHeaders.put(httpHeader.name().asString(), httpHeader.value().asString()));
+                pushHandler.doPromisedRequest(promisedStreamId, promisedHeaders);
```

Comment:
Given that the threading was not identified as an issue, can we just pass `ListFW<HttpHeaderFW>` instead of `Map<String, String`> as the second argument to the push handler?